### PR TITLE
feat: implement command to fix block connections

### LINKED ISSUE #313: Connections Command
# Issue
Whenever you paste in a block with connections (Example: `//paste spruce_fence`, `//replace air iron_bars`), the block **does not have any connections**.

# Solution
I would like a `//update` command. This would update every single block in a region and **set connections** if it should have connections.

Alternate explanation:
```
foreach (x in region)
    foreach (y in region)
        foreach (z in region)
            if (block can have connections) update connections
        rof
    rof
rof
```
Need a 3rd one? From this:

to this:

# Extra
Possible arguments are `//update dir[south,west,north,east]` which could specify which directions are allowed.

Blocks that should have connections:
* Stairs
* Fences
* Walls
* Bars
* Glass

### PR DESCRIPTION
**Pull request by dordsor21**
- closes #313

### CODE REVIEW

**Comment by dordsor21** [comment]
more_fixies

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/processor/PlacementStateProcessor.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    private final MutableBlockVector3 clickedBlock = new MutableBlockVector3();
+
+    public PlacementStateProcessor(Extent extent, BlockTypeMask mask, boolean includeUnedited) {
+        super(extent);
+        // Required here as child classes are located within adapters and will therefore be statically accessed on startup,
+        // meaning we attempt to access BlockTypes class before it is correctly initialised.
+        if (!SETUP) {
+            synchronized (PlacementStateProcessor.class) {
+                if (!SETUP) {
+                    setup();
```
Comment (PierreSchwang) [inline worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/processor/PlacementStateProcessor.java:53]: `SETUP` should be set to `true` I guess?

**Comment by PierreSchwang** [comment]
The copyright header is missing in some files, no (e.g. FaweMutableBlockPlaceContext.java)? Though I'm wondering why the CI is not failing on that

**Comment by dordsor21** [comment]
> The copyright header is missing in some files, no (e.g. FaweMutableBlockPlaceContext.java)? Though I'm wondering why the CI is not failing on that

That should be because licences are only required in worldedit packages/classes

**Comment by AGuyNamedJens** [comment]
I couldn't wait for release (over 100k blocks.. not connected.. help) and can confirm, it works, however console does show an error `[17:58:33 WARN]: [com.fastasyncworldedit.core.queue.implementation.ParallelQueueExtent] Cannot invoke "net.minecraft.core.Holder.a()" because "this.E" is null
[17:58:33 ERROR]: [com.fastasyncworldedit.core.queue.implementation.ParallelQueueExtent] Catching
java.lang.NullPointerException: Cannot invoke "net.minecraft.core.Holder.a()" because "this.E" is null`

**Comment by dordsor21** [comment]
What's the full error?

**Comment by AGuyNamedJens** [comment]
`[18:02:23 ERROR]: [com.fastasyncworldedit.core.queue.implementation.ParallelQueueExtent] Catching
java.lang.NullPointerException: Cannot invoke "net.minecraft.core.Holder.a()" because "this.E" is null
        at net.minecraft.world.level.Level.dimensionType(Level.java:1796) ~[?:?]
        at net.minecraft.world.level.LevelReader.getMinBuildHeight(LevelReader.java:86) ~[?:?]
        at net.minecraft.world.level.LevelHeightAccessor.getMaxBuildHeight(LevelHeightAccessor.java:12) ~[purpur-1.20.4.jar:git-Purpur-2176]
        at net.minecraft.world.level.block.DoublePlantBlock.getStateForPlacement(DoublePlantBlock.java:53) ~[?:?]
        at com.sk89q.worldedit.bukkit.adapter.impl.fawe.v1_20_R3.PaperweightPlacementStateProcessor.getStateAtFor(PaperweightPlacementStateProcessor.java:63) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.extent.processor.PlacementStateProcessor.getBlockOrdinal(PlacementStateProcessor.java:436) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.extent.processor.PlacementStateProcessor.applyBlock(PlacementStateProcessor.java:452) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.extent.filter.LinkedFilter.applyBlock(LinkedFilter.java:28) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.extent.filter.block.CharFilterBlock.filter(CharFilterBlock.java:176) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.sk89q.worldedit.regions.Region.filter(Region.java:326) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.sk89q.worldedit.regions.Region.filter(Region.java:309) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.sk89q.worldedit.regions.CuboidRegion.filter(CuboidRegion.java:761) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.extent.filter.block.ChunkFilterBlock.filter(ChunkFilterBlock.java:88) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.queue.implementation.chunk.ChunkHolder.filterBlocks(ChunkHolder.java:914) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.queue.IQueueExtent.apply(IQueueExtent.java:167) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at com.fastasyncworldedit.core.queue.implementation.ParallelQueueExtent.lambda$apply$0(ParallelQueueExtent.java:178) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1403) ~[?:?]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
        at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]`

It appears to not support newer fences as in spruce_fence etc 
image

**Comment by dordsor21** [comment]
Fixed^

**Comment by VL4DST3R** [comment]
> Fixed^

Came here for both of the same reasons as the user above but encountered a different issue when trying to use it (it did not work). 

```
23:16:45 [SEVERE] An unexpected error while handling a FastAsyncWorldEdit command
23:16:45 java.lang.NullPointerException: Cannot invoke "com.sk89q.worldedit.world.block.BlockType.getInternalId()" because "type" is null
23:16:45 	at com.sk89q.worldedit.function.mask.BlockTypeMask.add(BlockTypeMask.java:105) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.fastasyncworldedit.core.extent.processor.PlacementStateProcessor.setup(PlacementStateProcessor.java:164) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.fastasyncworldedit.core.extent.processor.PlacementStateProcessor.<init>(PlacementStateProcessor.java:87) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.bukkit.adapter.impl.fawe.v1_20_R1.PaperweightPlacementStateProcessor.<init>(PaperweightPlacementStateProcessor.java:31) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.bukkit.adapter.impl.fawe.v1_20_R1.PaperweightFaweAdapter.getPlatformPlacementProcessor(PaperweightFaweAdapter.java:617) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.bukkit.BukkitServerInterface.getPlatformPlacementProcessor(BukkitServerInterface.java:319) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.command.RegionCommands.fixblocks(RegionCommands.java:914) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.command.RegionCommandsRegistration.cmd$_fixblocks(RegionCommandsRegistration.java:1034) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at org.enginehub.piston.CommandManager.execute(CommandManager.java:157) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:45 	at com.sk89q.worldedit.extension.platform.PlatformCommandManager.lambda$handleCommandOnCurrentThread$16(PlatformCommandManager.java:727) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.sk89q.worldedit.extension.platform.PlatformCommandManager.handleCommandTask(PlatformCommandManager.java:748) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.sk89q.worldedit.extension.platform.PlatformCommandManager.handleCommandOnCurrentThread(PlatformCommandManager.java:729) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.sk89q.worldedit.extension.platform.PlatformCommandManager.lambda$handleCommand$14(PlatformCommandManager.java:700) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.sk89q.worldedit.extension.platform.AbstractPlayerActor.lambda$runAction$1(AbstractPlayerActor.java:672) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$run$0(AsyncNotifyKeyedQueue.java:48) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$call$1(AsyncNotifyKeyedQueue.java:58) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
23:16:46 	at com.fastasyncworldedit.core.util.task.KeyQueuedExecutorService$KeyRunner.lambda$run$0(KeyQueuedExecutorService.java:150) ~[FastAsyncWorldEdit-Bukkit-2.11.1-SNAPSHOT.jar:?]
23:16:46 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(ForkJoinTask.java:1726) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(ForkJoinTask.java:1717) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinTask$InterruptibleTask.exec(ForkJoinTask.java:1641) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035) ~[?:?]
23:16:46 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187) ~[?:?]
```

Is this there a build of this available to download somewhere?

**Review by IronApollo — APPROVED** [review]
As always, looks good to me with positive feedback provided by test users. Good work on a long-awaited feature.

**Review by PierreSchwang — APPROVED** [review]

**Comment by VL4DST3R** [comment]
What is the command for this? I see it's been merged but I cannot find it under `//fixblocks`.

**Pull request merged**